### PR TITLE
feat: :sparkles: add Shodan InternetDB boefje

### PR DESCRIPTION
**Pull request by zcrt**
### Changes
Adds [Shodan InternetDB](https://internetdb.shodan.io) boefje and normalizer.

Unfortunately, InternetDB does not return a protocol for its ports. In addition, it is not competely clear for me how to use the DNSPTRRecord object.

### Demo
image

image

image

image

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [ ] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have made corresponding changes to the documentation, if necessary.
- [x] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

**Comment by underdarknl** [comment]
Frequently Asked Questions

    How much does it cost?
    It's free for non-commercial use! If you're using the InternetDB API to make money then you need an [enterprise license](https://enterprise.shodan.io/).
    How often is it updated?
    The API gets updated once a week.
    Do I need a Shodan API key?
    No, you don't need to have a Shodan account or a Shodan API key in order to use the InternetDB API.
    How is it different than the regular Shodan API?
    The InternetDB API provides a lot less data than a regular Shodan IP lookup. It only gives you the list of open ports, vulnerabilities, cpes, hostnames and tags. The InternetDB API doesn't return any banners. And it is only updated once a week whereas the regular Shodan API is updated continuously in real-time.

**Comment by underdarknl** [comment]
N.b. This API does not seem to have any way of ingesting an API-key even if you had one.

**Comment by zcrt** [comment]
Correct, see https://github.com/minvws/nl-kat-coordination/blob/ca16a41c209186877cf038289dbb9c7a1332643d/boefjes/boefjes/plugins/kat_internetdb/description.md?plain=1#L2

**Comment by zcrt** [comment]
There is probably something wrong with the feeding of DNSPTRRecords, how should that be formatted?

image

**Review by ammar92 — COMMENTED** [review]
Looks good to me, very minimal remarks. Can be tested

File: boefjes/boefjes/plugins/kat_internetdb/normalize.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from octopoes.models.ooi.dns.records import DNSPTRRecord
+from octopoes.models.ooi.dns.zone import Hostname
+from octopoes.models.ooi.findings import CVEFindingType, Finding
+from octopoes.models.ooi.network import IPPort, Network, PortState
+from octopoes.models.ooi.software import Software, SoftwareInstance
+
+DNS_PTR_STR = ".in-addr.arpa"
+
+
+def cpe_to_name_version(cpe: str) -> tuple:
```
Comment (dekkers) [inline boefjes/boefjes/plugins/kat_internetdb/normalize.py:16]: There is also a `get_name_from_cpe` in kat_binaryedge/services/normalize.py. I think it would be better to have a general cpe parsing function in plugins/helpers.py that is used by both this normalizer and the binaryedge normalizers. Functions like this should also have unit tests to make sure all edge cases are parsed correctly.

**Review by dekkers — COMMENTED** [review]
This boefje should be renamed to "Shodan InternetDB" to make clear it is for querying the shodan database.

File: octopoes/octopoes/models/ooi/network.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ class IPPort(OOI):
     object_type: Literal["IPPort"] = "IPPort"
 
     address: Reference = ReferenceField(IPAddress, max_issue_scan_level=0, max_inherit_scan_level=4)
-    protocol: Protocol
+    protocol: Protocol | None = None
```
[the remark this answers is not in the document]
Comment (zcrt) [reply octopoes/octopoes/models/ooi/network.py:84]: While I do agree that an IP port should specify its protocol, Shodan InternetDB returns both TCP and UDP in a single list, without specifying the protocol

File: boefjes/boefjes/plugins/kat_internetdb/main.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+REQUEST_TIMEOUT = 60
+
+
+def run(boefje_meta: BoefjeMeta) -> list[tuple[set, bytes | str]]:
+    """Make request to InternetDB."""
+    ip = boefje_meta.arguments["input"]["address"]
+    results = {}
+    if ip_address(ip).is_private:
+        logging.info("Private IP requested, I will not forward this to Shodan InternetDB.")
```
[the remark this answers is not in the document]
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_internetdb/main.py:17]: This is how the current Shodan boefje also handles the request, should that be changed as well?

File: boefjes/boefjes/plugins/kat_internetdb/main.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    else:
+        try:
+            results = requests.get(f"https://internetdb.shodan.io/{ip}", timeout=REQUEST_TIMEOUT).json()
+        except requests.exceptions.RequestException as exc:
+            logging.warning(exc)
+
+    return [(set(), json.dumps(results))]
```
[the remark this answers is not in the document]
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_internetdb/main.py:24]: Moving the `json` conversion means the boefje succeeds if the request returns an invalid/non-json response. That means the normalizer fails in that case, is that your intended behavior?

File: boefjes/boefjes/plugins/kat_internetdb/normalize.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        logging.info("No InternetDB results available for normalization.")
+    elif "detail" in result:
+        if result["detail"] == "No information available":
+            logging.info("No information available for IP.")
+        else:
+            logging.warning("Unexpected detail: %s", str(result["detail"]))
+    else:
+        if result["ip"] != input_ooi_str:
+            logging.warning("Returned IP different from input OOI.")
+            logging.debug("Result IP: %s, Input IP: %s)", result["ip"], input_ooi_str)
```
[the remark this answers is not in the document]
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_internetdb/normalize.py:42]: Agreed, however, one might not want to log explicit IP-addresses in a production system. That is why that is added as a separate debug statement

File: boefjes/boefjes/plugins/kat_internetdb/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_internetdb/main.py:17]: I noticed that this will still result in the execution of the normalizer, how would one prevent that?

File: boefjes/boefjes/plugins/kat_internetdb/normalize.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        for hostname in result["hostnames"]:
+            hostname_ooi = Hostname(name=hostname, network=Network(name=input_["network"]["name"]).reference)
+            yield hostname_ooi
+            if hostname.endswith(DNS_PTR_STR):
```
[the remark this answers is not in the document]
Comment (dekkers) [reply boefjes/boefjes/plugins/kat_internetdb/normalize.py]: It should reference the target hostname. For example if example.org is the reverse DNS of 192.0.2.1, then we have a PTR of 1.2.0.192.in-addr.arpa. with example.org as data. The hostname it should reference is example.org, not 1.2.0.192.in-addr.arpa.

File: boefjes/boefjes/plugins/kat_internetdb/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dekkers) [reply boefjes/boefjes/plugins/kat_internetdb/main.py]: As far as  I can see there currently isn't a way to do, it looks like only raising an exception will cause it to not add the default raw file type for the normalizer.

[the remark this answers is not in the document]
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_internetdb/main.py]: Raised in a separate issue: https://github.com/minvws/nl-kat-coordination/issues/3054

File: boefjes/boefjes/plugins/kat_internetdb/normalize.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_internetdb/normalize.py]: The target hostname does not have to be a single hostname and can even be nonexistent. After discussion with @underdarknl we agreed that the `.in-addr` is the preferred hostname for this object.

**Comment by zcrt** [comment]
It is unclear to me what current code change causes the robot test to fail: https://github.com/minvws/nl-kat-coordination/actions/runs/9465358509/job/26074702650?pr=2615

**Comment by zcrt** [comment]
Seems to be fixed by merging from main ¯\_(ツ)_/¯

**Comment by madelondohmen** [comment]
Outcome of discussion meeting:
Remove IPPort from yielded objects. Changes in the model are therefore not needed, making the PR smaller. Additionally, include a warning in the description of the boefje that potentially outdated information remains when the boefje is disabled.

_(See https://github.com/minvws/nl-kat-coordination/discussions/3173 for full discussion)_

**Comment by zcrt** [comment]
>  Remove IPPort from yielded objects. Changes in the model are therefore not needed, making the PR smaller. Additionally, include a warning in the description of the boefje that potentially outdated information remains when the boefje is disabled.

A smaller PR does not make sense in this context. InternetDB yields useful information about ports, it unfortunately omits the protocol. The normalizer can create a port object with protocol `None`. The Shodan/Nmap/whatever normalizer can update this object with the correct protocol, isn't this how KAT is supposed to work?

**Comment by underdarknl** [comment]
> > Remove IPPort from yielded objects. Changes in the model are therefore not needed, making the PR smaller. Additionally, include a warning in the description of the boefje that potentially outdated information remains when the boefje is disabled.
> 
> A smaller PR does not make sense in this context. InternetDB yields useful information about ports, it unfortunately omits the protocol. The normalizer can create a port object with protocol `None`. The Shodan/Nmap/whatever normalizer can update this object with the correct protocol, isn't this how KAT is supposed to work?

The Protocol is Part of the Primary key of the IpPort, because we need to be able to track Both a TCP and a UDP port on port 53 (for example).
If you where to add Just the port 53 on an IP, you'd end up with an object that in itself contains useful information, eg, on this IP port 53 is open. If you where then to add the protocol by running Nmap-TCP OpenKAT would need to decide to upgrade the IpPort to include the TCP protocol, and possible later also create the UDP version. However, since both of those are different objects, and have different proofs from the partial data supplied by InternetDB we (for now) end up making new objects Next to the InternetDB IpPort. 
Consolidation of partial information by enriching it with extra information from other tools is not too hard when you know that some information will always be missing for some specific tool, or if you know that a specific tools output is always superseded by information from a different tool but it gets complicated fast once you get two or more tools that send out conflicting information, or information that is not meant to be consolidated (we had no protocol, there is no one single correct consolidation, as both TCP and UDP are valid "upgrades", which do you pick? Do they overwrite each other? Does InternetDB know for a fact that the third option (no protocol) is the valid fact, and are there in fact three different valid objects that should exists?

We do agree (partially) that having a notion of an Open Port is better than not knowing that the port is open, especially when no l2 tools can be ran, or when no other source has port-informaton that includes the protocol. However, that would mean that you'd need to make assumptions in our bits later on. Is a port-443 or port-80 an issue? Its listed as allowed for TCP, but since we dont know which protocol we are dealing with, its hard to say for sure if its an issue. 

The absence of this information and the issues that follow from this are closely related to the work we have been doing with regards to Octopoesv3. Partial information, consolidation, level of accuracy of tools, age of the information we are consolidating, etc.

**Comment by zcrt** [comment]
> However, since both of those are different objects, and have different proofs from the partial data supplied by InternetDB we (for now) end up making new objects Next to the InternetDB IpPort.

Can you elaborate on this? Since a port without protocol makes little sense, it means we are dealing with missing information that should be enhanced by other plugins, instead of separate objects. Of course, it _could_ be the case that InternetDB meant UDP 53 at point X in time and Nmap found TCP 53 at point Y in time. A simple algorithm may conclude that InternetDB also meant TCP 53. This would even be more likely rationally, since Nmap is a higher-level boefje. The likelihood of this edge-case occurring however, should decrease with increasing the amount of sources and  refresh interval of those sources.

> The absence of this information and the issues that follow from this are closely related to the work we have been doing with regards to Octopoesv3. Partial information, consolidation, level of accuracy of tools, age of the information we are consolidating, etc.

Indeed, I think it is essential to have a robust model to deal with uncertainty, as KAT is build around the idea of fetching information from a lot of different sources, which almost inevitably will result in conflicting information at some point. My suggestion would be to explore using confidence scores for this.

> However, that would mean that you'd need to make assumptions in our bits later on. Is a port-443 or port-80 an issue? 

I think this decision should be made by the sysadmin that knows which ports should be open and which should be closed, instead of a hardcoded default list in OpenKAT. All open ports are interesting to some extent, some ports are usually just more interesting than others. Thus, OpenKAT may very well assist by adding confidence scores to common and uncommon ports. If the protocol is not known this would add some uncertainty.

File: boefjes/boefjes/plugins/kat_shodan_internetdb/cover.jpg
Comment (underdarknl) [inline boefjes/boefjes/plugins/kat_shodan_internetdb/cover.jpg:1]: this needs compressing.

File: boefjes/boefjes/plugins/kat_shodan_internetdb/normalize.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            logging.warning("Unexpected detail: %s", result["detail"])
+    else:
+        for hostname in result["hostnames"]:
+            hostname_ooi = Hostname(name=hostname, network=Network(name=input_ooi["network"]["name"]).reference)
+            yield hostname_ooi
+            if hostname.endswith(DNS_PTR_STR):
+                yield DNSPTRRecord(hostname=hostname_ooi.reference, value=hostname, address=input_ooi_reference)
+
+        for port in result["ports"]:
+            yield IPPort(address=input_ooi_reference, port=int(port), state=PortState("open"))
```
Comment (underdarknl) [inline boefjes/boefjes/plugins/kat_shodan_internetdb/normalize.py:37]:
```suggestion
        #for port in result["ports"]:
        #    yield IPPort(address=input_ooi_reference, port=int(port), state=PortState("open"))
```

File: octopoes/octopoes/models/ooi/network.py (hunk 2 above)
Comment (underdarknl) [inline octopoes/octopoes/models/ooi/network.py:84]:
```suggestion
    protocol: Protocol
```

File: boefjes/boefjes/plugins/kat_shodan_internetdb/cover.jpg
Comment (zcrt) [reply boefjes/boefjes/plugins/kat_shodan_internetdb/cover.jpg:1]: Is the acceptable size defined somewhere?

Comment (underdarknl) [reply boefjes/boefjes/plugins/kat_shodan_internetdb/cover.jpg:1]: https://github.com/minvws/nl-kat-coordination/issues/1904
Not a definition yet, but on its way to become a CI check. @HeleenSG what size would you think we want to limit to?

Comment (HeleenSG) [reply boefjes/boefjes/plugins/kat_shodan_internetdb/cover.jpg:1]: I'll look into it.

**Review by underdarknl — APPROVED** [review]
